lib/model/model.go

```diff
@@ -1151,46 +1278,94 @@ func (m *model) ClusterConfig(deviceID protocol.DeviceID, cm protocol.ClusterCon
 		m.Add(is)
 	}
 
-	m.pmut.Lock()
-	m.remotePausedFolders[deviceID] = paused
-	m.pmut.Unlock()
+	return changed, tempIndexFolders, paused, nil
+}
 
-	// This breaks if we send multiple CM messages during the same connection.
-	if len(tempIndexFolders) > 0 {
-		m.pmut.RLock()
-		conn, ok := m.conn[deviceID]
-		m.pmut.RUnlock()
-		// In case we've got ClusterConfig, and the connection disappeared
-		// from infront of our nose.
-		if ok {
-			m.progressEmitter.temporaryIndexSubscribe(conn, tempIndexFolders)
-		}
+func (m *model) ccCheckEncryption(fcfg config.FolderConfiguration, folderDevice config.FolderDeviceConfiguration, ccDeviceRemote, ccDeviceLocal protocol.Device, hasDeviceRemote, hasDeviceLocal, deviceUntrusted bool) error {
+	hasTokenRemote := hasDeviceRemote && len(ccDeviceRemote.EncPwToken) > 0
+	hasTokenLocal := hasDeviceLocal && len(ccDeviceLocal.EncPwToken) > 0
+	isEncRemote := folderDevice.EncryptionPassword != ""
+	isEncLocal := fcfg.Type == config.FolderTypeReceiveEncrypted
+
+	if !isEncRemote && !isEncLocal && deviceUntrusted {
+		return errEncNotEncryptedUntrusted
 	}
 
-	if deviceCfg.Introducer {
-		folders, devices, foldersDevices, introduced := m.handleIntroductions(deviceCfg, cm)
-		folders, devices, deintroduced := m.handleDeintroductions(deviceCfg, foldersDevices, folders, devices)
-		if introduced || deintroduced {
-			changed = true
-			cfg := m.cfg.RawCopy()
-			cfg.Folders = make([]config.FolderConfiguration, 0, len(folders))
-			for _, fcfg := range folders {
-				cfg.Folders = append(cfg.Folders, fcfg)
-			}
-			cfg.Devices = make([]config.DeviceConfiguration, len(devices))
-			for _, dcfg := range devices {
-				cfg.Devices = append(cfg.Devices, dcfg)
-			}
-			m.cfg.Replace(cfg)
-		}
+	if !(hasTokenRemote || hasTokenLocal || isEncRemote || isEncLocal) {
+		// Noone cares about encryption here
+		return nil
 	}
 
-	if changed {
-		if err := m.cfg.Save(); err != nil {
-			l.Warnln("Failed to save config", err)
+	if isEncRemote && isEncLocal {
+		// Should never happen, but config racyness and be safe.
+		return errEncInvConfigLocal
```

Comment:
I guess I am missing how this works, but I thought both sides have to agree on encryption?

Reply:
Still bad naming: `isEncryptionRemote` means "locally unencrypted data will be encrypted for this remote device" while `isEncryptedLocal` means "the local data is already encrypted". The two are mutually exclusive, because you can't (well shouldn't) encrypt again locally encrypted data.